Source/Core/InputCommon/ControllerInterface/OctagonalMouseJail.cpp

```diff
@@ -0,0 +1,413 @@
+      static_cast<long>(
+          m_center_of_render_window.second -
+          (fraction_of_screen_to_lock_mouse_in_y * percentage_reduction_for_corner_gates))};
+
+  octagon_points[EAST] = std::pair<PrimaryPrimitive, PrimaryPrimitive>{
+      static_cast<long>(
+          std::ceil(m_center_of_render_window.first + fraction_of_screen_to_lock_mouse_in_x)),
+      m_center_of_render_window.second};
+
+  octagon_points[SOUTH_EAST] = std::pair<PrimaryPrimitive, PrimaryPrimitive>{
+      static_cast<long>(m_center_of_render_window.first + (fraction_of_screen_to_lock_mouse_in_x *
+                                                           percentage_reduction_for_corner_gates)),
+      static_cast<long>(
+          m_center_of_render_window.second +
+          (fraction_of_screen_to_lock_mouse_in_y * percentage_reduction_for_corner_gates))};
+
+  octagon_points[SOUTH] = std::pair<PrimaryPrimitive, PrimaryPrimitive>{
+      m_center_of_render_window.first,
+      static_cast<long>(
+          std::ceil(m_center_of_render_window.second + fraction_of_screen_to_lock_mouse_in_y))};
+
+  octagon_points[SOUTH_WEST] = std::pair<PrimaryPrimitive, PrimaryPrimitive>{
+      static_cast<long>(m_center_of_render_window.first - (fraction_of_screen_to_lock_mouse_in_x *
+                                                           percentage_reduction_for_corner_gates)),
+      static_cast<long>(
+          m_center_of_render_window.second +
+          (fraction_of_screen_to_lock_mouse_in_y * percentage_reduction_for_corner_gates))};
+
+  octagon_points[WEST] = std::pair<PrimaryPrimitive, PrimaryPrimitive>{
+      static_cast<long>(
+          std::floor(m_center_of_render_window.first - fraction_of_screen_to_lock_mouse_in_x)),
+      m_center_of_render_window.second};
+
+  octagon_points[NORTH_WEST] = std::pair<PrimaryPrimitive, PrimaryPrimitive>{
+      static_cast<long>(m_center_of_render_window.first - (fraction_of_screen_to_lock_mouse_in_x *
+                                                           percentage_reduction_for_corner_gates)),
+      static_cast<long>(
+          m_center_of_render_window.second -
+          (fraction_of_screen_to_lock_mouse_in_y * percentage_reduction_for_corner_gates))};
+  return octagon_points;
+}
+
+bool OctagonalMouseJail::IsPointInsideOctagon(
+    const std::pair<PrimaryPrimitive, PrimaryPrimitive>& mouse_point,
+    const std::array<std::pair<PrimaryPrimitive, PrimaryPrimitive>, 8>& octagon_points)
+{
+  //  This function uses a raycasting method to figure out if a point is  inside the octagon
+  //  casting a ray horizontally in one direction from the mouse point and if the
+  //  casted ray intersects with the edges of the octagon an even number of times (0 counts)
+  //  the point is outside of the octagon. If the ray intersects an odd number of times, the
+  //  ray is inside of the polygon. Jordan Curve Thereom is the name I was told
+  //  https://erich.realtimerendering.com/ptinpoly/
```

Comment:
You're better off using the [winding number](https://en.wikipedia.org/wiki/Winding_number) here, it's a more stable algorithm for checking if a point is inside a polygon.

See also https://web.archive.org/web/20170511054708/http://geomalgorithms.com/a03-_inclusion.html

Reply:
Am I required to use winding number? Not a huge deal if I do, it's just a lot of work to recreate something that already works. My algorithm is stable from my experience playing Melee. I haven't had a single instance where the algorithm failed.

Reply:
It's probably fine as-is -- your polygons are probably more well behaved compared to something that goes through transformation matrices for typical game collision detection -- but I wanted to note it anyway.